### PR TITLE
Nemotron-H: MLP gate cleanup

### PR DESCRIPTION
**Pull request by DominguesM**
Summary:

- MLP cleanup for Nemotron-H: pass NULL for unused gate tensors.

Changes:

- MLP gate cleanup: `src/llama-model.cpp:14337` — in `llm_build_nemotronh::build_ffn_layer`, pass `NULL` for gate tensors to make it explicit the MLP has no gate.

Deferred (out of scope here):

- Hybrid GGUF metadata, block_type array, and hybrid caches plumbing.
- Causal Conv1D padding tweaks beyond the default `d_conv - 1` (kept as-is).
- MLP activation as an hparam (Nemotron-H remains `LLM_FFN_RELU_SQR`).

Status:

- Converts and loads; generation works. Sample output below.

Sample generation

```
main: prompt: '<SPECIAL_10>System

<SPECIAL_11>User
/no_think Write a haiku about GPUs
<SPECIAL_11>Assistant
<think></think>'
main: number of tokens in prompt = 26
     1 -> '<s>'
    10 -> '<SPECIAL_10>'
  6775 -> 'System'
  1267 -> '

'
    11 -> '<SPECIAL_11>'
  4019 -> 'User'
  1010 -> '
'
 75010 -> '/no'
 31180 -> '_th'
  2077 -> 'ink'
 27040 -> ' Write'
  1261 -> ' a'
  3944 -> ' ha'
 15636 -> 'iku'
  2314 -> ' about'
 20534 -> ' GP'
  7176 -> 'Us'
  1010 -> '
'
    11 -> '<SPECIAL_11>'
102897 -> 'Assistant'
  1010 -> '
'
 49250 -> '<th'
  2077 -> 'ink'
  4468 -> '></'
 74045 -> 'think'
  1062 -> '>'

sampler seed: 3146942522
sampler params: 
	repeat_last_n = 64, repeat_penalty = 1,000, frequency_penalty = 0,000, presence_penalty = 0,000
	dry_multiplier = 0,000, dry_base = 1,750, dry_allowed_length = 2, dry_penalty_last_n = 4096
	top_k = 40, top_p = 0,950, min_p = 0,050, xtc_probability = 0,000, xtc_threshold = 0,100, typical_p = 1,000, top_n_sigma = -1,000, temp = 0,800
	mirostat = 0, mirostat_lr = 0,100, mirostat_ent = 5,000
sampler chain: logits -> logit-bias -> penalties -> dry -> top-n-sigma -> top-k -> typical -> top-p -> min-p -> xtc -> temp-ext -> dist 
generate: n_ctx = 4096, n_batch = 2048, n_predict = 128, n_keep = 1

System

User
/no_think Write a haiku about GPUs
Assistant
<think></think>Silicon hearts beat,
Light-speed math in parallel,
GPUs dream in pixels.
 [end of text]
```

Addresses https://github.com/ggml-org/llama.cpp/issues/15409 - Feature Request: Support for NVidia Nemotron Nano v2

### CODE REVIEW

File: src/llama-model.cpp [hunk 1]
```diff
@@ -11151,7 +11210,8 @@ struct llm_graph_context_mamba : public llm_graph_context {
         ggml_tensor * ssm_states_all  = mctx_cur->get_s_l(il);
 
         ggml_tensor * conv = build_rs(inp, conv_states_all, hparams.n_embd_r(), n_seqs);
-        conv = ggml_reshape_3d(ctx0, conv, d_conv - 1, d_inner, n_seqs);
+        const int64_t conv_pad = hparams.ssm_conv_padding;
```
Comment (gabe-l-hart) [inline src/llama-model.cpp]: This is interesting. I've got the model working with batch-size 1 without the need for conv padding. Is it possible that this is only required for batch size > 1? If so, it would be great to get this PR rebuilt on top of my latest branch to only include the changes needed for this padding.

File: src/llama-model.cpp [hunk 2]
```diff
@@ -14182,18 +14243,28 @@ struct llm_build_nemotronh : public llm_graph_context_mamba {
         for (int il = 0; il < n_layer; ++il) {
             struct ggml_tensor * inpSA = inpL;
 
-            // norm
-            cur = build_norm(inpL,
-                    model.layers[il].attn_norm, NULL,
-                    LLM_NORM_RMS, il);
+            ggml_tensor * norm_w = hparams.is_recurrent(il) || hparams.n_ff(il) == 0
```
Comment (gabe-l-hart) [inline src/llama-model.cpp]: Nice find. This was the last fix I needed on my branch to get it working.

File: src/llama-model.cpp [hunk 3]
```diff
@@ -14271,26 +14341,27 @@ struct llm_build_nemotronh : public llm_graph_context_mamba {
...
         const llama_model & model,
         const int           il) {
 
-        ggml_tensor * ffn_inp = ggml_add(ctx0, cur, inpSA);
-        cb(ffn_inp, "ffn_inp", il);
-        cur = build_norm(ffn_inp,
-                model.layers[il].ffn_norm, NULL,
-                LLM_NORM_RMS, il);
-                cb(cur, "ffn_norm", il);
+        llm_ffn_op_type act = LLM_FFN_SILU;
```
Comment (gabe-l-hart) [inline src/llama-model.cpp]: Nice. I opted to simply use `LLM_FFN_RELU_SQR` for this model assuming that that is a static part of the architecture. That removed the need to plumb this through as an `hparam`. I think that's preferable unless/until a model is released with the same architecture but a different MLP activation choice.

File: src/llama-model.cpp [hunk 4]
```diff
@@ -14271,26 +14341,27 @@ struct llm_build_nemotronh : public llm_graph_context_mamba {
...
+        if (hparams.mlp_act == LLAMA_MLP_ACT_RELU_SQR) {
+            act = LLM_FFN_RELU_SQR;
+        }
 
         cur = build_ffn(cur,
                 model.layers[il].ffn_up,   model.layers[il].ffn_up_b,   NULL,
-                model.layers[il].ffn_gate, model.layers[il].ffn_gate_b, NULL,
+                NULL,                      NULL,                         NULL,
```
Comment (gabe-l-hart) [inline src/llama-model.cpp]: This is technically a no-op since those tensor pointers will be `NULL` anyway. It's a bit cleaner this way, though, since it makes it clear that this model does not use a gate in the MLP blocks. This would be a nice cleanup piece to include in an updated version of the PR to get your name on the final merge if you'd like to.

File: convert_hf_to_gguf.py [hunk 5]
```diff
@@ -7577,31 +7577,146 @@ def __init__(self, *args, **kwargs):
...
+        self.gguf_writer.add_uint32(f"{self.gguf_writer.arch}.mamba.head_dim", self.hparams["mamba_head_dim"])
+        self.gguf_writer.add_float32(f"{self.gguf_writer.arch}.mamba.time_step_min", self.hparams["time_step_min"])
+        self.gguf_writer.add_float32(f"{self.gguf_writer.arch}.mamba.time_step_max", self.hparams["time_step_max"])
+        self.gguf_writer.add_float32(f"{self.gguf_writer.arch}.mamba.time_step_floor", self.hparams["time_step_floor"])
+        self.gguf_writer.add_array(
+            f"{self.gguf_writer.arch}.mamba.time_step_limit",
+            [float(x) for x in self.hparams["time_step_limit"]],
+        )
+        # Depthwise conv is causal: stride=1, padding=kernel-1
+        self.gguf_writer.add_ssm_conv_padding(self.conv_kernel - 1)
```
Comment (gabe-l-hart) [inline convert_hf_to_gguf.py]: It looks like the padding is still `d_conv - 1`, so let's hold off adding the padding `hparam` for now.

File: src/llama-model.cpp (hunk 1 above)
Comment (gabe-l-hart) [reply src/llama-model.cpp]: Looking at the conversion script, it looks like this odel still uses `d_conv - 1` for the padding, so let's not add this now.

File: src/llama-model.cpp [hunk 6]
```diff
@@ -3228,7 +3287,7 @@ bool llama_model::load_tensors(llama_model_loader & ml) {
                     const uint32_t v_dim              = head_dim;
                     const int64_t num_attention_heads = hparams.n_head();
                     const int64_t q_num_heads         = num_attention_heads;
-                    const int64_t dt_dim              = std::max(64, int(hparams.n_embd / 16));
+                    const int64_t dt_dim              = hparams.ssm_dt_rank > 0 ? hparams.ssm_dt_rank : std::max<int64_t>(64, hparams.n_embd / 16);
```
Comment (gabe-l-hart) [inline src/llama-model.cpp]: This one is interesting. In the `config.json` for the model, the `time_step_rank` is `256`. `n_embd` will be `4480`, so `n_embd / 16 = 280 (!= 256)`. This also means that this would kick in the `std::max` and set `dt_dim` to `64`. This means that having this change in does tangibly change the `hparams`, but I'm not fully sure what difference that would make to the model's performance. I'll look a bit to see if this is important.

**Review by gabe-l-hart — CHANGES_REQUESTED** [review]
Thank you for putting this together! I've stumbled on a bunch of the same fixes as well. I'd love to get your name on the final merge, so there are two things I think we could keep from this PR to help:

1. Some basic cleanup (eg. using `NULL` for unused gate tensors in the MLP)
2. _Possibly_ the inclusion of the plumbing to use `ssm_dt_rank`. I need to test this one more to see what difference it actually makes.

File: src/llama-model.cpp (hunk 6 above)
Comment (gabe-l-hart) [reply src/llama-model.cpp]: Also, I just noticed that this is in the `PLAMO2` case, so doesn't seem to be in the path for `NEMOTRONH` at all.

File: src/llama-model.cpp [hunk 7]
```diff
@@ -17211,7 +17283,7 @@ struct llm_build_plamo2 : public llm_graph_context_mamba {
             cb(x_bcdt, "mamba_bcdt_proj", il);
 
             // split into dt, B, C
-            const int64_t dt_dim = std::max(64, int(hparams.n_embd / 16));
+            const int64_t dt_dim = hparams.ssm_dt_rank > 0 ? hparams.ssm_dt_rank : std::max<int64_t>(64, hparams.n_embd / 16);
```
Comment (gabe-l-hart) [inline src/llama-model.cpp]: This also appears to only be in the `plamo2` case. This might be a separate PR if this has a tangible impact there?

**Comment by DominguesM** [comment]
> Thank you for putting this together! I've stumbled on a bunch of the same fixes as well. I'd love to get your name on the final merge, so there are two things I think we could keep from this PR to help:
> 
>     1. Some basic cleanup (eg. using `NULL` for unused gate tensors in the MLP)
> 
>     2. _Possibly_ the inclusion of the plumbing to use `ssm_dt_rank`. I need to test this one more to see what difference it actually makes.

Perfect, I'll do it

File: src/llama-model.cpp (hunk 6 above)
Comment (DominguesM) [reply src/llama-model.cpp]: On dt_dim: I set it to honor ssm_dt_rank (GGUF time_step_rank) when > 0, otherwise fall back to max(64, n_embd/16) (see src/llama-model.cpp:3256 and src/llama-model.cpp:17268). For Nemotron-H, time_step_rank=256 while n_embd/16=280; the fallback would be 64, so honoring the provided rank does change the effective hparams. I’m happy to keep it as-is, restrict it to plamo2, or move it to a separate PR—whichever you prefer.

With these minimal changes (no conv padding tweaks, MLP gate cleanup, optional dt_dim from rank), generation runs.

File: src/llama-model.cpp [hunk 8]
```diff
@@ -3253,7 +3253,9 @@ bool llama_model::load_tensors(llama_model_loader & ml) {
                     const uint32_t v_dim              = head_dim;
                     const int64_t num_attention_heads = hparams.n_head();
                     const int64_t q_num_heads         = num_attention_heads;
-                    const int64_t dt_dim              = std::max(64, int(hparams.n_embd / 16));
+                    const int64_t dt_dim              = hparams.ssm_dt_rank > 0
```
Comment (gabe-l-hart) [inline src/llama-model.cpp]: Since this is in the `LLM_ARCH_PLAMO2` case, I think this is not actually getting stimulated in this PR. I'd be curious if you made the same change for the `nemotronh` model what the result would be.

File: src/llama-model.cpp [hunk 9]
```diff
@@ -14339,11 +14341,11 @@ struct llm_build_nemotronh : public llm_graph_context_mamba {
 
         cur = build_ffn(cur,
                 model.layers[il].ffn_up,   model.layers[il].ffn_up_b,   NULL,
-                model.layers[il].ffn_gate, model.layers[il].ffn_gate_b, NULL,
+                NULL,                      NULL,                        NULL,
                 model.layers[il].ffn_down, model.layers[il].ffn_down_b, NULL,
                 NULL,
                 LLM_FFN_RELU_SQR, LLM_FFN_PAR, il);
-                cb(cur, "ffn_out", il);
+        cb(cur, "ffn_out", il);
```
Comment (gabe-l-hart) [inline src/llama-model.cpp:14346]: Nice catch

File: src/llama-model.cpp [hunk 10]
```diff
@@ -17263,7 +17265,9 @@ struct llm_build_plamo2 : public llm_graph_context_mamba {
             cb(x_bcdt, "mamba_bcdt_proj", il);
 
             // split into dt, B, C
-            const int64_t dt_dim = std::max(64, int(hparams.n_embd / 16));
+            const int64_t dt_dim = hparams.ssm_dt_rank > 0
```
Comment (gabe-l-hart) [inline src/llama-model.cpp]: Similarly, this is in `build_plamo2_mamba_layer`, so not taking effect for `nemotronh`. Can you try moving this over and see what happens?

**Review by gabe-l-hart — CHANGES_REQUESTED** [review]
I don't think we should modify `PLAMO2` in this PR, but I'm curious how those `dt_rank` changes would change the result for `nemotronh` if added there.

File: src/llama-model.cpp (hunk 10 above)
Comment (DominguesM) [reply src/llama-model.cpp]: I reverted the PLaMo‑2 dt_dim change to the default max(64, n_embd/16) for now (see src/llama-model.cpp:3256 and src/llama-model.cpp:17266). I’ll test the ssm_dt_rank variant separately and, if it looks beneficial, open a focused PR.

**Review by gabe-l-hart — APPROVED** [review]
Approving the cleanups for now. Thanks!
